keps/sig-cli/3515-kubectl-explain-openapiv3/README.md

```diff
@@ -0,0 +1,982 @@
+https://testgrid.k8s.io/sig-testing-canaries#ci-kubernetes-coverage-unit
+
+This can inform certain test coverage improvements that we want to do before
+extending the production code to implement this enhancement.
+-->
+
+- `k8s.io/kubectl/pkg/explain`: `09/29/2022`-`75.6`
+
+##### Integration tests
+
+<!--
+This question should be filled when targeting a release.
+For Alpha, describe what tests will be added to ensure proper quality of the enhancement.
+
+For Beta and GA, add links to added tests together with links to k8s-triage for those tests:
+https://storage.googleapis.com/k8s-triage/index.html
+-->
+
+- <test>: <link to test coverage>
+
+Tests should include
+
+- Expected Output tests
+- Show correct OpenAPI v3 endpoints are hit
+- Tests that show default/nullability information is being included in plaintext output
+- Tests that update the backing openapi in between calls to explain
+
+##### e2e tests
+
+<!--
+This question should be filled when targeting a release.
+For Alpha, describe what tests will be added to ensure proper quality of the enhancement.
+
+For Beta and GA, add links to added tests together with links to k8s-triage for those tests:
+https://storage.googleapis.com/k8s-triage/index.html
+
+We expect no non-infra related flakes in the last month as a GA graduation criteria.
+-->
+
+Existing e2e tests should be adapted for the new system.
+E2E test that shows every definition in OpenAPI document can be retrieved via explain
+
+
+- <test>: <link to test coverage>
+
+### Graduation Criteria
+
+Defined using feature gate
+
+#### Alpha
+
+- Feature implemented behind a command line flag `--experimental-output`
+- `--experimental-output` flag added
+- Existing explain tests are working or adapted for new implementation
+- Plaintext output roughly matches explain output
+- OpenAPIV3 (raw json) output implemented
+- HTML and MD outputs are not target for alpha
+
+#### Beta
+
+- md output implemented (or dropped from design due to continued debate)
+  - Table of contents all GVKs grouped by Group then Version.
+  - Section for each individual GVK
+  - All types hyperlink to specific section
+- basic html output  (or dropped from design due to continued debate)
+  - Table of contents all GVKs grouped by Group then Version.
+  - Page for each individual GVK.
+  - All types hyperlink to their specific page
+  - Searchable by name, description, field name.
+- kube-openAPI v3 JSON deserialization is optimized to take less than 150ms on
+  most machines
+- OpenAPI V3 is enabled by default on at least one version within kubectl's support window
+- Experimental flag is removed/made on by default (thus openapi v3 will always be tried first)
+- Old `kubectl explain` implementation for OpenAPI v2 remains as a fallback if v3 is unavailable
+(this policy should stand only until kubectl's version skew includes apiserver versions which enabled OpenAPI V3 by default). 
+
+#### GA
+
+- `--experimental-output` renamed to `--output`
```

Comment:
Note that anything we promote to [beta needs to stick around for one release after deprecation](https://kubernetes.io/docs/reference/using-api/deprecation-policy/#deprecating-a-flag-or-cli). Perhaps another SIG CLI maintainer can correct me on this, but that sounds like we'd want to switch the flag name at beta, not GA, so we don't have to go through a deprecation cycle on the old name if the feature successfully graduates. Aaaand if we rename the flag at beta graduation, it might make sense to have a graduation criterion that all planned formats have been present in at least one alpha release, instead of allowing them to be introduced at beta. Alternatively, any new formats being introduced while the basic formats graduate should be explicitly marked experimental themselves, decoupling their graduation from that of the openapiv3 migration in general.

Reply:
That is a good point. I'm open to multiple alpha cycles. Ill update the KEP